File: src/main/java/org/magicghostvu/actorvt/behavior/Behaviors.java
package org.magicghostvu.actorvt.behavior;

import org.magicghostvu.actorvt.context.GeneralActorContext;
import org.magicghostvu.actorvt.context.timer.TimerManData;

public final class Behaviors {

    private Behaviors() {
    }

    @SuppressWarnings("unchecked")
    public static <T> Behavior<T> same() {
        return (Behavior<T>) Same.INSTANCE;
    }

    @SuppressWarnings("unchecked")
    public static <T> Behavior<T> stopped() {
        return (Behavior<T>) Stopped.INSTANCE;
    }

    public static <T> Behavior<T> setUp(ActorVTFunc<GeneralActorContext<T>, Behavior<T>> factory) {
        return new SetUpBehavior<>(factory);
    }

    public static <T> Behavior<T> withTimer(ActorVTFunc<TimerManData<T>, Behavior<T>> timerFunc) {
        return new TimerBehavior<>(timerFunc);
    }
}

File: src/main/kotlin/org/magicghostvu/actorvt/behavior/Behaviors.kt
package org.magicghostvu.actorvt.behavior

import org.magicghostvu.actorvt.context.GeneralActorContext
import org.magicghostvu.actorvt.context.timer.TimerManData

object Behaviors {
    @JvmStatic
    fun <T> same(): Behavior<T> {
        return Same as Behavior<T>
    }


    @JvmStatic
    fun <T> stopped(): Behavior<T> {
        return Stopped as Behavior<T>
    }

    @JvmStatic
    public fun <T> setUp(factory: (GeneralActorContext<T>) -> Behavior<T>): Behavior<T> {
        return SetUpBehavior(factory)
    }


    @JvmStatic
    public fun <T> withTimer(setupTimer: (TimerManData<T>) -> Behavior<T>): Behavior<T> {
        return TimerBehavior(setupTimer)
    }


}

internal data object Same : Behavior<Any>()
internal data object Stopped : Behavior<Any>()